inferlo/forney/optimization/nfg_map_lp.py

```diff
@@ -0,0 +1,98 @@
+from collections import namedtuple
+from itertools import product
+import cvxpy as cp
+import numpy as np
+
+
+from inferlo.forney.nfg_model import NormalFactorGraphModel
+
+map_lp_result = namedtuple('map_lp_result', ['upper_bound',
+                                             'lower_bound',
+                                             'factor_beliefs',
+                                             'variable_beliefs',
+                                             'normalization_duals',
+                                             'marginalization_duals'
+                                             ])
+
+
+def map_lp(model: NormalFactorGraphModel) -> map_lp_result:
+    """MAP linear programming relaxation for
+        pairwise model with finite alphabet
+    """
+    al_size = model._default_domain.size()
+    var_size = len(model.edges)
+
+    factor_list = model.factors
+
+    variable_beliefs = cp.Variable((var_size, al_size), nonneg=True)
+    factor_beliefs = []
+    for f in range(len(factor_list)):
+        curr_factor_beliefs = {}
+        var_vals = list(product(range(al_size),
+                                repeat=len(factor_list[f].var_idx)))
+        for x in var_vals:
+            curr_factor_beliefs[x] = cp.Variable(nonneg=True)
+        factor_beliefs.append(curr_factor_beliefs)
+    obj = 0
+    cons = []
+
+    # add objective
+    for f in range(len(factor_list)):
+        var_vals = list(product(range(al_size),
+                                repeat=len(factor_list[f].var_idx)))
+        for x in var_vals:
+            value = factor_list[f].value(list(x))
+            if (value != 0):
+                obj += np.log(value) * factor_beliefs[f][x]
+
+    # normalization constraints
```

Comment:
Please format comments as sentences (capitalize and end with a period).
See https://google.github.io/styleguide/pyguide.html#386-punctuation-spelling-and-grammar